erizo/src/erizo/media/ExternalOutput.cpp

```diff
@@ -369,62 +409,69 @@ int ExternalOutput::deliverEvent_(MediaEventPtr event) {
 }
 
 bool ExternalOutput::initContext() {
-  if (video_codec_ != AV_CODEC_ID_NONE &&
-            audio_codec_ != AV_CODEC_ID_NONE &&
-            video_stream_ == nullptr &&
-            audio_stream_ == nullptr) {
-    AVCodec* video_codec = avcodec_find_encoder(video_codec_);
-    if (video_codec == nullptr) {
-      ELOG_ERROR("Could not find video codec");
-      return false;
-    }
+  if (input_video_codec_id_ != AV_CODEC_ID_NONE &&
+            input_audio_codec_id_ != AV_CODEC_ID_NONE &&
+            initialized_context_ == false) {
     need_to_send_fir_ = true;
     video_queue_.setTimebase(video_map_.clock_rate);
-    video_stream_ = avformat_new_stream(context_, video_codec);
-    video_stream_->id = 0;
-    video_stream_->codec->codec_id = video_codec_;
-    video_stream_->codec->width = 640;
-    video_stream_->codec->height = 480;
-    video_stream_->time_base = (AVRational) { 1, 30 };
-    video_stream_->metadata = genVideoMetadata();
-    // A decent guess here suffices; if processing the file with ffmpeg,
-    // use -vsync 0 to force it not to duplicate frames.
-    video_stream_->codec->pix_fmt = PIX_FMT_YUV420P;
-    if (context_->oformat->flags & AVFMT_GLOBALHEADER) {
-      video_stream_->codec->flags |= CODEC_FLAG_GLOBAL_HEADER;
-    }
-    context_->oformat->flags |= AVFMT_VARIABLE_FPS;
+    if (avio_open(&context_->pb, context_->filename, AVIO_FLAG_WRITE) < 0) {
+      ELOG_ERROR("Error opening output context.");
+      exit(1);
+    } else {
+      AVCodecID output_video_codec_id = this->bestMatchOutputCodecId(input_video_codec_id_);
+      AVCodecID output_audio_codec_id = this->bestMatchOutputCodecId(input_audio_codec_id_);
 
-    AVCodec* audio_codec = avcodec_find_encoder(audio_codec_);
-    if (audio_codec == nullptr) {
-      ELOG_ERROR("Could not find audio codec");
-      return false;
-    }
+      if (input_video_codec_id_ == output_video_codec_id) {
+        ELOG_INFO("Video input codec matches output codec, don't need to transcode.");
+      } else {
+        need_video_transcode_ = true;
+      }
+      if (input_audio_codec_id_ == output_audio_codec_id) {
+        ELOG_INFO("Audio input codec matches output codec, don't need to transcode.");
+      } else {
+        need_audio_transcode_ = true;
+      }
 
-    audio_stream_ = avformat_new_stream(context_, audio_codec);
-    audio_stream_->id = 1;
-    audio_stream_->codec->codec_id = audio_codec_;
-    audio_stream_->codec->sample_rate = audio_map_.clock_rate;
-    audio_stream_->time_base = (AVRational) { 1, audio_stream_->codec->sample_rate };
-    audio_stream_->codec->channels = audio_map_.channels;
-    if (context_->oformat->flags & AVFMT_GLOBALHEADER) {
-      audio_stream_->codec->flags |= CODEC_FLAG_GLOBAL_HEADER;
-    }
+      video_decoder_.initDecoder(input_video_codec_id_,
```

Comment:
We don't need to initialize the decoders unless we're transcoding, right?

Reply:
Yes, that's right we should be able to avoid initialize decoders in some cases.
I didn't want to go deeper here until get things done and define surrounding API's, since didn't look as much resource waste, for example, in the case of Output MKV, it seems to only increase ~15mb of memory when you are recording.